src/query_plan/operation.rs

```diff
@@ -0,0 +1,426 @@
+        let normalized_selections = normalize_selections(&selection_set.selections, fragments);
+        NormalizedSelectionSet {
+            ty: selection_set.ty.clone(),
+            selections: normalized_selections,
+        }
+    }
+}
+
+impl From<NormalizedSelectionSet> for SelectionSet {
+    fn from(val: NormalizedSelectionSet) -> Self {
+        SelectionSet {
+            ty: val.ty.clone(),
+            selections: flatten_selections(&val.selections),
+        }
+    }
+}
+
+/// Converts vec of Selections to a map of NormalizedSelections
+///
+/// Expands all named fragments and merge duplicate selections.
+fn normalize_selections(
+    selections: &Vec<Selection>,
+    fragments: &IndexMap<Name, Node<Fragment>>,
+) -> IndexMap<String, NormalizedSelection> {
+    let mut normalized: IndexMap<String, NormalizedSelection> = IndexMap::new();
+    for selection in selections {
+        match selection {
+            Selection::Field(field) => {
+                let expanded_selection_set =
+                    normalize_selections(&field.selection_set.selections.to_owned(), fragments);
+                let key = selection_field_key(field);
+
+                if let NormalizedSelection::NormalizedField(field_entry) =
+                    normalized.entry(key.to_owned()).or_insert(
+                        NormalizedSelection::NormalizedField(Node::new(NormalizedField {
+                            definition: field.definition.clone(),
+                            alias: field.alias.clone(),
+                            name: field.name.clone(),
+                            arguments: field.arguments.clone(),
+                            directives: field.directives.clone(),
+                            selection_set: NormalizedSelectionSet {
+                                ty: field.selection_set.ty.clone(),
+                                selections: IndexMap::new(),
+                            },
+                        })),
+                    )
+                {
+                    let merged_selections = merge_selections(
+                        &field_entry.selection_set.selections,
+                        &expanded_selection_set,
+                    );
+                    let mut selection_set = field_entry.selection_set.clone();
+                    selection_set.selections = merged_selections;
+                    field_entry.make_mut().selection_set = selection_set;
+                    // field_entry.selection_set.selections = merged_selections;
+                }
+            }
+            Selection::FragmentSpread(named_fragment) => {
+                if let Some(fragment) = fragments.get(&named_fragment.fragment_name) {
+                    let expanded_selection_set = normalize_selections(
+                        &fragment.selection_set.selections.to_owned(),
+                        fragments,
+                    );
+                    normalized = merge_selections(&normalized, &expanded_selection_set);
```

Comment:
This normalizes fragment spreads to the normalization of the fragment definition’s selection set. Shouldn’t it create an inline fragment instead, in order to preserve the type condition and directives?

Reply:
good callout, I'll double check with the JS implementation (but regardless we should still be converting to the inline fragments due to the above)

Reply:
@SimonSapin is right here. The relevant code is in [these lines](https://github.com/apollographql/federation/blob/7903a296a96948d96f37e2638a6f3e165a413d28/internals-js/src/operations.ts#L3618-L3620), which gets called by `expandAllFragments()` when dealing with named fragment spreads.

The gist is that if both of the following are true:
1. The named fragment's type equals the parent type of the fragment spread.
2. The named fragment's fragment spread and fragment definition have no directive applications.

Then we can effectively treat the named fragment's selection set's selections as top-level selections, and merge them into `normalized`. Otherwise, we create an inline fragment. (More generally, what's happening here is that the JS codebase is "rebasing" the inline fragment onto its parent as an optimization, since in this case it's cheap computationally and it's guaranteed to succeed).

Two things to note though. The first is that `normalize_selections()` doesn't currently take in any argument for `parent_type`; you'll need to add that argument, and provide it at the callsites from the selection sets (alternatively, it looks like all the current callsites are effectively taking their selections from selection sets, so you could just pass a selection set reference into the function and rename it to e.g. `normalize_selection_set()`).

The second thing is that you might be surprised about point 2, where we oddly look at the fragment definition's directive applications in addition to the fragment spread's. This is part of the JS codebase; if you look at the code [here](https://github.com/apollographql/federation/blob/7903a296a96948d96f37e2638a6f3e165a413d28/internals-js/src/operations.ts#L3472), you'll notice that the directives for what JS codebase calls the "operation element" of the fragment spread consists of the concatenation of the named fragment's definition's directive applications and the spread's directive applications. This accordingly affects the JS codebase's `key` for the fragment spread (but doesn't affect its `toString()`).

I'm tempted to write this off as a bug for now, and to only look at the fragment spread's directive applications. There aren't any spec directives (both in GraphQL spec or the federation spec) that go on the `FRAGMENT_DEFINITION` location. (The other option would be to add a field to `NormalizedSelectionKey::InlineFragment` for storing `named_fragment_directives`, but then if we ever converted the operation to the `apollo-rs` representation or to a string and converted back, that data would be gone.)